licensing/notice.go

```diff
@@ -0,0 +1,283 @@
+// under the License.
+
+package licensing
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"go/build"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+	"text/tabwriter"
+	"text/template"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/sirkon/goproxy/gomod"
+	"gopkg.in/src-d/go-license-detector.v2/licensedb"
+)
+
+// DefaultNoticeHeader is used as the default NOTICE header.
+const DefaultNoticeHeader = `{{.Project}}
+Copyright {{.ProjectYears}} {{.Licensor}}
+
+This product includes software developed at {{.Licensor}} and
+third-party software developed by the licenses listed below.
+`
+
+// NoticeFormat is the template format for the generated notice file.
+const NoticeFormat = `%s
+=========================================================================
+
+{{.DependencyBlob}}
+=========================================================================
+`
+
+// Notice contains all of the information
+type Notice struct {
+	// Licensor is the source code's owner / maintainer.
+	Licensor string
+
+	// Project is the source code's project or repository name.
+	Project string
+
+	// ProjectYears contains the years that the project has been running
+	// in the {{start year}}-{{current year}} or {{year}} format depending
+	// if it's a single or multi-year effort.
+	ProjectYears string
+
+	// Dependencies contains the dependency list that can be consumed in a
+	// programatic way.
+	Dependencies []Dependency
+
+	// DependencyBlob is a stringified version of Dependencies that has been
+	// processed through the tabwritter package so the columns are aligned.
+	// Only populated when Writer is not nil.
+	DependencyBlob string
+}
+
+// Dependency contains the dependency name and the license.
+type Dependency struct {
+	Name, License string
+}
+
+// GenerateNoticeParams is consumed by GenerateNotice.
+type GenerateNoticeParams struct {
+	// GoModFile is the location of the `go.mod` file to open.
+	GoModFile string
+
+	// Licensor of the codebase.
+	Licensor string
+
+	// Project name.
+	Project string
+
+	// Project's start year so it can be reflected in the NOTICE file.
+	StartYear int
+
+	// Writer where to write the NOTICE templated output. If not specified,
+	// writing the template will be skipped, and Notice won't have a populated
+	// DependencyBlob.
+	Writer io.Writer
+
+	// NoticeHeader overrides the DefaultNoticeHeader used in NOTICE. Only relevant
+	// when Writer is not nil.
+	NoticeHeader string
+
+	// AnalyseFunc returns a list of results with their license.
+	AnalyseFunc func(args ...string) []licensedb.Result
+}
+
+// Validate ensures that the structure is usable by its consumer.
+func (params GenerateNoticeParams) Validate() error {
+	var merr = new(multierror.Error)
+	if params.GoModFile == "" {
+		merr = multierror.Append(merr, errors.New("notice: missing file path"))
+	}
+	if params.AnalyseFunc == nil {
+		merr = multierror.Append(merr, errors.New("notice: AnalyseFunc is missing"))
```

Comment:
`nit` and completely optional. This message is not following the pattern of the other two 😄 🤓